modules/dnn/src/cuda4dnn/csl/cudnn/cudnn.hpp

```diff
@@ -0,0 +1,201 @@
+                    );
+                } else {
+                    std::vector<int> stride(rank);
+                    stride.back() = 1;
+                    /* WHAT WE HAVE NOW:
+                     * stride[-1] = 1
+                     * stride[-2] = garbage
+                     * stride[-3] = garbage
+                     * stride[-4] = garbage
+                     * ...
+                     */
+
+                    std::copy(start + 1, end, stride.begin());
+                    /* WHAT WE HAVE NOW:
+                     * stride[-1] = 1
+                     * stride[-2] = dim[-1]
+                     * stride[-3] = dim[-2]
+                     * stride[-4] = dim[-3]
+                     * ...
+                     */
+
+                    std::partial_sum(stride.rbegin(), stride.rend(), stride.rbegin(), std::multiplies<int>());
+                    /* WHAT WE HAVE NOW:
+                     * stride[-1] = 1
+                     * stride[-2] = stride[-1] * dim[-1]
+                     * stride[-3] = stride[-2] * dim[-2]
+                     * stride[-4] = stride[-3] * dim[-3]
+                     * ...
+                     */
+
+                    std::vector<int> dims(start, end);
+                    CUDA4DNN_CHECK_CUDNN(
+                        cudnnSetTensorNdDescriptor(descriptor,
+                            detail::get_data_type<T>(), rank,
+                            dims.data(), stride.data()
+                        )
+                    );
+                }
+            } catch (...) {
+                /* cudnnDestroyTensorDescriptor will not fail */
+                CUDA4DNN_CHECK_CUDNN(cudnnDestroyTensorDescriptor(descriptor));
+                throw;
+            }
+        }
+
+        cudnnTensorDescriptor_t descriptor;
+    };
+
+    /** @brief element-wise addition with broadcasting
+     *
+     * \f$ C = \alpha A + \beta C \f$
+     *
+     * @tparam          T           matrix element type (must be `half` or `float`)
+     *
+     * @param           handle      valid cuDNN handle
+     * @param           alpha       scale factor for A
+     * @param           aDesc       tensor descriptor for A
+     * @param[in]       A           pointer to tensor in device memory
+     * @param           beta        scale factor for C
+     * @param           cDesc       tensor descriptor for C
+     * @param[in]       C           pointer to tensor in device memory
+     *
+     * Exception Guarantee: Basic
+     */
+    template <class T>
+    typename std::enable_if<std::is_same<T, half>::value || std::is_same<T, float>::value, void>
+    ::type add(const Handle& handle,
+        T alpha, const TensorDescriptor<T>& aDesc, DevicePtr<const T> A,
+        T beta, const TensorDescriptor<T>& cDesc, DevicePtr<T> C)
+    {
+        CUDA4DNN_CHECK_CUDNN(
+            cudnnAddTensor(HandleAccessor::get(handle),
+                &alpha, aDesc.get(), A.get(),
+                &beta, cDesc.get(), C.get()
```

Comment:
[BUG] should specalize for `half` type since `cudnnAddTensor` expects `float` scaling factors for both `half` and `float`.

Reference: https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#scaling-parameters